In [ ]:
%run 02-connect.ipynb

In [2]:
%%sql

SELECT *
FROM aapl_historical_stock_price
LIMIT 10;

date    year  month    open    high     low   close      volume  id
0  "1/30/14"  2014.0    1.0  502.54  506.50  496.70  499.78  24182996.0   1
1  "1/29/14"  2014.0    1.0  503.95  507.37  498.62  500.75  17991828.0   2
2  "1/28/14"  2014.0    1.0  508.76  515.00  502.07  506.50  38119083.0   3
3  "1/27/14"  2014.0    1.0  550.07  554.80  545.75  550.50  20602736.0   4
4  "1/24/14"  2014.0    1.0  554.00  555.62  544.75  546.07  15483491.0   5
5  "1/23/14"  2014.0    1.0  549.94  556.50  544.81  556.18  14425478.0   6
6  "1/22/14"  2014.0    1.0  550.91  557.29  547.81  551.51  13602762.0   7
7  "1/21/14"  2014.0    1.0  540.99  550.07  540.42  549.07  11750792.0   8
8  "1/17/14"  2014.0    1.0  551.48  552.07  539.90  540.67  15489527.0   9
9  "1/16/14"  2014.0    1.0  554.90  556.85  551.68  554.25   8210190.0  10

## COUNT

In [2]:
%%sql

SELECT COUNT(*)
FROM aapl_historical_stock_price;

COUNT(*)
0      3555

In [3]:
%%sql

SELECT COUNT(high)
FROM aapl_historical_stock_price;

COUNT(high)
0         3555

In [4]:
%%sql

SELECT COUNT(date)
FROM aapl_historical_stock_price;

COUNT(date)
0         3555

In [5]:
%%sql

SELECT COUNT(date) AS count_of_date
FROM aapl_historical_stock_price;

count_of_date
0           3555

In [6]:
%%sql

SELECT COUNT(date) AS "Count Of Date"
FROM aapl_historical_stock_price;

Count Of Date
0           3555

## SUM

In [3]:
%%sql

SELECT SUM(volume)
FROM aapl_historical_stock_price;

SUM(volume)
0  7.358540e+10

## SQL MIN/MAX

In [4]:
%%sql

SELECT MIN(volume) AS min_volume,
       MAX(volume) AS max_volume
FROM aapl_historical_stock_price;

min_volume   max_volume
0         0.0  120466292.0

## SQL AVG

In [5]:
%%sql

SELECT AVG(high)
FROM aapl_historical_stock_price;

AVG(high)
0  165.314031

## SQL GROUP BY

In [6]:
%%sql

SELECT year,
       COUNT(*) AS count
FROM aapl_historical_stock_price
GROUP BY year;

year  count
0   2014.0     20
1   2013.0    252
2   2012.0    250
3   2011.0    252
4   2010.0    259
5   2009.0    257
6   2008.0    255
7   2007.0    251
8   2006.0    251
9   2005.0    252
10  2004.0    252
11  2003.0    252
12  2002.0    252
13  2001.0    248
14  2000.0    252

In [7]:
%%sql

SELECT year,
       month,
       COUNT(*) AS count
FROM aapl_historical_stock_price
GROUP BY year, month;

year  month  count
0    2014.0    1.0     20
1    2013.0   12.0     21
2    2013.0   11.0     20
3    2013.0   10.0     23
4    2013.0    9.0     20
..      ...    ...    ...
164  2000.0    5.0     22
165  2000.0    4.0     19
166  2000.0    3.0     23
167  2000.0    2.0     20
168  2000.0    1.0     20

[169 rows x 3 columns]

In [8]:
%%sql

SELECT year,
       month,
       COUNT(*) AS count
  FROM aapl_historical_stock_price
 GROUP BY 1, 2;

year  month  count
0    2014.0    1.0     20
1    2013.0   12.0     21
2    2013.0   11.0     20
3    2013.0   10.0     23
4    2013.0    9.0     20
..      ...    ...    ...
164  2000.0    5.0     22
165  2000.0    4.0     19
166  2000.0    3.0     23
167  2000.0    2.0     20
168  2000.0    1.0     20

[169 rows x 3 columns]

In [9]:
%%sql

SELECT year,
       month,
       COUNT(*) AS count
FROM aapl_historical_stock_price
GROUP BY year, month
ORDER BY month, year;

year  month  count
0    2000.0    1.0     20
1    2001.0    1.0     21
2    2002.0    1.0     21
3    2003.0    1.0     21
4    2004.0    1.0     20
..      ...    ...    ...
164  2009.0   12.0     23
165  2010.0   12.0     22
166  2011.0   12.0     21
167  2012.0   12.0     20
168  2013.0   12.0     21

[169 rows x 3 columns]

## SQL HAVING

In [11]:
%%sql

SELECT year,
       month,
       MAX(high) AS month_high
FROM aapl_historical_stock_price
GROUP BY year, month
HAVING MAX(high) > 400
ORDER BY year, month
LIMIT 10;

year  month  month_high
0  2011.0    7.0      404.50
1  2011.0    9.0      422.86
2  2011.0   10.0      426.70
3  2011.0   11.0      408.00
4  2011.0   12.0      409.09
5  2012.0    1.0      458.24
6  2012.0    2.0      547.61
7  2012.0    3.0      621.45
8  2012.0    4.0      644.00
9  2012.0    5.0      596.76

## SQL CASE

In [15]:
%%sql

SELECT player_name,
    year,
    CASE
        WHEN year = '"SR"' THEN 'yes'
        ELSE NULL
    END AS is_a_senior
FROM college_football_players
LIMIT 10;

player_name  year is_a_senior
0  "Ralph Abernathy"  "JR"        None
1     "Mekale McKay"  "SO"        None
2      "Trenier Orr"  "SO"        None
3     "Bennie Coney"  "FR"        None
4    "Johnny Holton"  "JR"        None
5    "Howard Wilder"  "JR"        None
6   "Munchie Legaux"  "SR"         yes
7        "Mark Barr"  "FR"        None
8      "Aaron Brown"  "FR"        None
9  "Anthony McClung"  "SR"         yes

In [18]:
%%sql

SELECT player_name,
    year,
    CASE
        WHEN year = '"SR"' THEN 'yes'
        ELSE 'no'
    END AS is_a_senior
FROM college_football_players
LIMIT 10;

player_name  year is_a_senior
0  "Ralph Abernathy"  "JR"          no
1     "Mekale McKay"  "SO"          no
2      "Trenier Orr"  "SO"          no
3     "Bennie Coney"  "FR"          no
4    "Johnny Holton"  "JR"          no
5    "Howard Wilder"  "JR"          no
6   "Munchie Legaux"  "SR"         yes
7        "Mark Barr"  "FR"          no
8      "Aaron Brown"  "FR"          no
9  "Anthony McClung"  "SR"         yes

In [19]:
%%sql

SELECT player_name,
    weight,
    CASE
        WHEN weight > 250 THEN 'over 250'
        WHEN weight > 200 THEN '201-250'
        WHEN weight > 175 THEN '176-200'
        ELSE '175 or under'
    END AS weight_group
FROM college_football_players
LIMIT 10;

player_name  weight  weight_group
0  "Ralph Abernathy"   161.0  175 or under
1     "Mekale McKay"   195.0       176-200
2      "Trenier Orr"   177.0       176-200
3     "Bennie Coney"   216.0       201-250
4    "Johnny Holton"   190.0       176-200
5    "Howard Wilder"   180.0       176-200
6   "Munchie Legaux"   200.0       176-200
7        "Mark Barr"   163.0  175 or under
8      "Aaron Brown"   172.0  175 or under
9  "Anthony McClung"   177.0       176-200

While the above works, it's really best practice to create statements that don't overlap. WHEN weight > 250 and WHEN weight > 200 overlap for every value greater than 250, which is a little confusing. A better way to write the above would be:

In [20]:
%%sql

SELECT player_name,
    weight,
    CASE
        WHEN weight > 250 THEN 'over 250'
        WHEN weight > 200
        AND weight <= 250 THEN '201-250'
        WHEN weight > 175
        AND weight <= 200 THEN '176-200'
        ELSE '175 or under'
    END AS weight_group
FROM college_football_players
LIMIT 10;

player_name  weight  weight_group
0  "Ralph Abernathy"   161.0  175 or under
1     "Mekale McKay"   195.0       176-200
2      "Trenier Orr"   177.0       176-200
3     "Bennie Coney"   216.0       201-250
4    "Johnny Holton"   190.0       176-200
5    "Howard Wilder"   180.0       176-200
6   "Munchie Legaux"   200.0       176-200
7        "Mark Barr"   163.0  175 or under
8      "Aaron Brown"   172.0  175 or under
9  "Anthony McClung"   177.0       176-200

In [22]:
%%sql

SELECT player_name,
    CASE
        WHEN year = '"FR"'
        AND position = '"WR"' THEN 'frosh_wr'
        ELSE NULL
    END AS sample_case_statement
FROM college_football_players
LIMIT 10;

player_name sample_case_statement
0  "Ralph Abernathy"                  None
1     "Mekale McKay"                  None
2      "Trenier Orr"                  None
3     "Bennie Coney"                  None
4    "Johnny Holton"                  None
5    "Howard Wilder"                  None
6   "Munchie Legaux"                  None
7        "Mark Barr"              frosh_wr
8      "Aaron Brown"                  None
9  "Anthony McClung"                  None

CASE's slightly more complicated and substantially more useful functionality comes from pairing it with aggregate functions. For example, let's say you want to only count rows that fulfill a certain condition. Since COUNT ignores nulls, you could use a CASE statement to evaluate the condition and produce null or non-null values depending on the outcome:

In [24]:
%%sql

SELECT CASE
        WHEN year = '"FR"' THEN 'FR'
        ELSE 'Not FR'
    END AS year_group,
    COUNT(1) AS count
FROM college_football_players
GROUP BY CASE
        WHEN year = '"FR"' THEN 'FR'
        ELSE 'Not FR'
    END;

year_group  count
0     Not FR  16633
1         FR   9665

Now, you might be thinking "why wouldn't I just use a WHERE clause to filter out the rows I don't want to count?" You could do that—it would look like this:

In [26]:
%%sql

SELECT COUNT(1) AS fr_count
FROM college_football_players
WHERE year = '"FR"';

fr_count
0      9665

But what if you also wanted to count a couple other conditions? Using the WHERE clause only allows you to count one condition. Here's an example of counting multiple conditions in one query:

In [27]:
%%sql

SELECT CASE WHEN year = '"FR"' THEN 'FR'
            WHEN year = '"SO"' THEN 'SO'
            WHEN year = '"JR"' THEN 'JR'
            WHEN year = '"SR"' THEN 'SR'
            ELSE 'No Year Data' END AS year_group,
            COUNT(1) AS count
  FROM college_football_players
 GROUP BY 1;

year_group  count
0            JR   5665
1            SO   5881
2            FR   9665
3            SR   5086
4  No Year Data      1

In some instances, you might want to show data horizontally. This is known as "pivoting" (like a pivot table in Excel):

In [30]:
%%sql

SELECT COUNT(
        CASE
            WHEN year = '"FR"' THEN 1
            ELSE NULL
        END
    ) AS fr_count,
    COUNT(
        CASE
            WHEN year = '"SO"' THEN 1
            ELSE NULL
        END
    ) AS so_count,
    COUNT(
        CASE
            WHEN year = '"JR"' THEN 1
            ELSE NULL
        END
    ) AS jr_count,
    COUNT(
        CASE
            WHEN year = '"SR"' THEN 1
            ELSE NULL
        END
    ) AS sr_count
FROM college_football_players

fr_count  so_count  jr_count  sr_count
0      9665      5881      5665      5086

## SQL DISTINCT

In [31]:
%%sql

SELECT DISTINCT month
FROM aapl_historical_stock_price
LIMIT 10

month
0    1.0
1   12.0
2   11.0
3   10.0
4    9.0
5    8.0
6    7.0
7    6.0
8    5.0
9    4.0

In [32]:
%%sql

SELECT DISTINCT year,
    month
FROM aapl_historical_stock_price
LIMIT 10

year  month
0  2014.0    1.0
1  2013.0   12.0
2  2013.0   11.0
3  2013.0   10.0
4  2013.0    9.0
5  2013.0    8.0
6  2013.0    7.0
7  2013.0    6.0
8  2013.0    5.0
9  2013.0    4.0

In [33]:
%%sql

SELECT COUNT(DISTINCT month) AS unique_months
FROM aapl_historical_stock_price

unique_months
0             12

## SQL JOINS

In [37]:
%%sql

SELECT teams.conference AS conference,
    AVG(players.weight) AS average_weight
FROM college_football_players players
    JOIN college_football_teams teams ON teams.school_name = players.school_name
GROUP BY teams.conference
ORDER BY AVG(players.weight) DESC
LIMIT 10

conference  average_weight
0                "SEC"      230.416970
1                "ACC"      228.995521
2            "Big Ten"      228.377899
3  "American Athletic"      227.979537
4             "Pac-12"      226.537400
5             "Big 12"      226.253782
6       "Mid-American"      225.721264
7     "Conference USA"      225.371906
8   "FBS Independents"      224.738815
9    "Missouri Valley"      224.525934

## SQL INNER JOIN

In [38]:
%%sql

SELECT players.*,
    teams.*
FROM college_football_players players
    JOIN college_football_teams teams ON teams.school_name = players.school_name
LIMIT 10

full_school_name   school_name        player_name position  height  \
0  "Cincinnati Bearcats"  "Cincinnati"  "Ralph Abernathy"     "RB"    67.0   
1  "Cincinnati Bearcats"  "Cincinnati"     "Mekale McKay"     "WR"    78.0   
2  "Cincinnati Bearcats"  "Cincinnati"      "Trenier Orr"     "CB"    71.0   
3  "Cincinnati Bearcats"  "Cincinnati"     "Bennie Coney"     "QB"    75.0   
4  "Cincinnati Bearcats"  "Cincinnati"    "Johnny Holton"     "WR"    75.0   
5  "Cincinnati Bearcats"  "Cincinnati"    "Howard Wilder"     "DB"    71.0   
6  "Cincinnati Bearcats"  "Cincinnati"   "Munchie Legaux"     "QB"    77.0   
7  "Cincinnati Bearcats"  "Cincinnati"        "Mark Barr"     "WR"    73.0   
8  "Cincinnati Bearcats"  "Cincinnati"      "Aaron Brown"     "CB"    71.0   
9  "Cincinnati Bearcats"  "Cincinnati"  "Anthony McClung"     "WR"    73.0   

   weight  year          hometown state  id                    division  \
0   161.0  "JR"          "ATLANTA   GA"   0  "FBS (Division I-A Teams)"   
1   195.0  "SO"       "LOUISVILLE   KY"   0  "FBS (Division I-A Teams)"   
2   177.0  "SO"    "WINTER GARDEN   FL"   0  "FBS (Division I-A Teams)"   
3   216.0  "FR"       "PLANT CITY   FL"   0  "FBS (Division I-A Teams)"   
4   190.0  "JR"            "MIAMI   FL"   0  "FBS (Division I-A Teams)"   
5   180.0  "JR"       "SEA ISLAND   GA"   0  "FBS (Division I-A Teams)"   
6   200.0  "SR"      "NEW ORLEANS   LA"   0  "FBS (Division I-A Teams)"   
7   163.0  "FR"  "FORT LAUDERDALE   FL"   0  "FBS (Division I-A Teams)"   
8   172.0  "FR"            "MIAMI   FL"   0  "FBS (Division I-A Teams)"   
9   177.0  "SR"     "INDIANAPOLIS   IN"   0  "FBS (Division I-A Teams)"   

            conference   school_name  \
0  "American Athletic"  "Cincinnati"   
1  "American Athletic"  "Cincinnati"   
2  "American Athletic"  "Cincinnati"   
3  "American Athletic"  "Cincinnati"   
4  "American Athletic"  "Cincinnati"   
5  "American Athletic"  "Cincinnati"   
6  "American Athletic"  "Cincinnati"   
7  "American Athletic"  "Cincinnati"   
8  "American Athletic"  "Cincinnati"   
9  "American Athletic"  "Cincinnati"   

                                          roster_url  id  
0  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1  
1  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1  
2  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1  
3  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1  
4  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1  
5  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1  
6  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1  
7  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1  
8  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1  
9  "http://espn.go.com/ncf/teams/roster?teamId=2132"   1

In [39]:
%%sql

SELECT players.school_name AS players_school_name,
    teams.school_name AS teams_school_name
FROM college_football_players players
    JOIN college_football_teams teams ON teams.school_name = players.school_name
LIMIT 10

players_school_name teams_school_name
0        "Cincinnati"      "Cincinnati"
1        "Cincinnati"      "Cincinnati"
2        "Cincinnati"      "Cincinnati"
3        "Cincinnati"      "Cincinnati"
4        "Cincinnati"      "Cincinnati"
5        "Cincinnati"      "Cincinnati"
6        "Cincinnati"      "Cincinnati"
7        "Cincinnati"      "Cincinnati"
8        "Cincinnati"      "Cincinnati"
9        "Cincinnati"      "Cincinnati"

## SQL LEFT JOIN

Let's start by running an INNER JOIN on the Crunchbase dataset and taking a look at the results. We'll just look at company-permalink in each table, as well as a couple other fields, to get a sense of what's actually being joined.

In [42]:
%%sql

SELECT companies.permalink AS companies_permalink,
    companies.name AS companies_name,
    acquisitions.company_permalink AS acquisitions_permalink,
    acquisitions.acquired_at AS acquired_date
FROM crunchbase_companies companies
    JOIN crunchbase_acquisitions acquisitions ON companies.permalink = acquisitions.company_permalink
LIMIT 10

companies_permalink       companies_name  \
0            "/company/waywire"           "#waywire"   
1       "/company/1000memories"       "1000memories"   
2          "/company/12society"          "12Society"   
3          "/company/280-north"          "280 North"   
4          "/company/280-north"          "280 North"   
5  "/company/2web-technologies"  "2Web Technologies"   
6              "/company/3leaf"              "3Leaf"   
7         "/company/3x-systems"         "3X Systems"   
8              "/company/4home"              "4Home"   
9               "/company/5to1"               "5to1"   

         acquisitions_permalink acquired_date  
0            "/company/waywire"    "10/17/13"  
1       "/company/1000memories"     "10/3/12"  
2          "/company/12society"      "7/3/13"  
3          "/company/280-north"      "7/1/10"  
4          "/company/280-north"      "7/1/10"  
5  "/company/2web-technologies"      "6/1/06"  
6              "/company/3leaf"     "2/19/11"  
7         "/company/3x-systems"    "10/25/12"  
8              "/company/4home"     "12/1/10"  
9               "/company/5to1"      "5/1/11"

You may notice that "280 North" appears twice in this list. That is because it has two entries in the tutorial.crunchbase_acquisitions table, both of which are being joined onto the tutorial.crunchbase_companies table.

Now try running that query as a LEFT JOIN:

In [43]:
%%sql

SELECT companies.permalink AS companies_permalink,
    companies.name AS companies_name,
    acquisitions.company_permalink AS acquisitions_permalink,
    acquisitions.acquired_at AS acquired_date
FROM crunchbase_companies companies
    LEFT JOIN crunchbase_acquisitions acquisitions ON companies.permalink = acquisitions.company_permalink
LIMIT 10

companies_permalink       companies_name  \
0            "/company/waywire"           "#waywire"   
1       "/company/1000memories"       "1000memories"   
2          "/company/12society"          "12Society"   
3          "/company/280-north"          "280 North"   
4          "/company/280-north"          "280 North"   
5  "/company/2web-technologies"  "2Web Technologies"   
6              "/company/3leaf"              "3Leaf"   
7         "/company/3x-systems"         "3X Systems"   
8              "/company/4home"              "4Home"   
9               "/company/5to1"               "5to1"   

         acquisitions_permalink acquired_date  
0            "/company/waywire"    "10/17/13"  
1       "/company/1000memories"     "10/3/12"  
2          "/company/12society"      "7/3/13"  
3          "/company/280-north"      "7/1/10"  
4          "/company/280-north"      "7/1/10"  
5  "/company/2web-technologies"      "6/1/06"  
6              "/company/3leaf"     "2/19/11"  
7         "/company/3x-systems"    "10/25/12"  
8              "/company/4home"     "12/1/10"  
9               "/company/5to1"      "5/1/11"

You can see that the first two companies from the previous result set, #waywire and 1000memories, are pushed down the page by a number of results that contain null values in the acquisitions_permalink and acquired_date fields.

This is because the LEFT JOIN command tells the database to return all rows in the table in the FROM clause, regardless of whether or not they have matches in the table in the LEFT JOIN clause.

## SQL RIGHT JOIN

RIGHT JOIN is rarely used because you can achieve the results of a RIGHT JOIN by simply switching the two joined table names in a LEFT JOIN. For example, in this query of the Crunchbase dataset, the LEFT JOIN section:

In [45]:
%%sql

SELECT companies.permalink AS companies_permalink,
    companies.name AS companies_name,
    company_permalink AS acquisitions_permalink,
    acquired_at AS acquired_date
FROM crunchbase_companies companies
    LEFT JOIN crunchbase_acquisitions acquisitions ON companies.permalink = company_permalink
LIMIT 10

companies_permalink       companies_name  \
0            "/company/waywire"           "#waywire"   
1       "/company/1000memories"       "1000memories"   
2          "/company/12society"          "12Society"   
3          "/company/280-north"          "280 North"   
4          "/company/280-north"          "280 North"   
5  "/company/2web-technologies"  "2Web Technologies"   
6              "/company/3leaf"              "3Leaf"   
7         "/company/3x-systems"         "3X Systems"   
8              "/company/4home"              "4Home"   
9               "/company/5to1"               "5to1"   

         acquisitions_permalink acquired_date  
0            "/company/waywire"    "10/17/13"  
1       "/company/1000memories"     "10/3/12"  
2          "/company/12society"      "7/3/13"  
3          "/company/280-north"      "7/1/10"  
4          "/company/280-north"      "7/1/10"  
5  "/company/2web-technologies"      "6/1/06"  
6              "/company/3leaf"     "2/19/11"  
7         "/company/3x-systems"    "10/25/12"  
8              "/company/4home"     "12/1/10"  
9               "/company/5to1"      "5/1/11"

produces the same results as this query:

In [46]:
%%sql

SELECT companies.permalink AS companies_permalink,
    companies.name AS companies_name,
    company_permalink AS acquisitions_permalink,
    acquired_at AS acquired_date
FROM crunchbase_acquisitions acquisitions
    RIGHT JOIN crunchbase_companies companies ON companies.permalink = company_permalink
LIMIT 10

companies_permalink       companies_name  \
0            "/company/waywire"           "#waywire"   
1       "/company/1000memories"       "1000memories"   
2          "/company/12society"          "12Society"   
3          "/company/280-north"          "280 North"   
4          "/company/280-north"          "280 North"   
5  "/company/2web-technologies"  "2Web Technologies"   
6              "/company/3leaf"              "3Leaf"   
7         "/company/3x-systems"         "3X Systems"   
8              "/company/4home"              "4Home"   
9               "/company/5to1"               "5to1"   

         acquisitions_permalink acquired_date  
0            "/company/waywire"    "10/17/13"  
1       "/company/1000memories"     "10/3/12"  
2          "/company/12society"      "7/3/13"  
3          "/company/280-north"      "7/1/10"  
4          "/company/280-north"      "7/1/10"  
5  "/company/2web-technologies"      "6/1/06"  
6              "/company/3leaf"     "2/19/11"  
7         "/company/3x-systems"    "10/25/12"  
8              "/company/4home"     "12/1/10"  
9               "/company/5to1"      "5/1/11"

In [48]:
%%sql

SELECT companies.permalink AS companies_permalink,
    companies.name AS companies_name,
    acquisitions.company_permalink AS acquisitions_permalink,
    acquisitions.acquired_at AS acquired_date
FROM crunchbase_companies companies
    LEFT JOIN crunchbase_acquisitions acquisitions ON companies.permalink = company_permalink
ORDER BY 1
LIMIT 10

companies_permalink       companies_name   acquisitions_permalink  \
0            "/company/0-6-com"            "0-6.com"                     None   
1             "/company/0xdata"             "0xdata"                     None   
2      "/company/1-800-doctors"      "1-800-DOCTORS"                     None   
3                  "/company/1"           "One Inc."                     None   
4        "/company/10-20-media"        "10-20 Media"                     None   
5  "/company/1000jobboersen-de"  "1000jobboersen.de"                     None   
6       "/company/1000memories"       "1000memories"  "/company/1000memories"   
7    "/company/1000museums-com"    "1000museums.com"                     None   
8         "/company/1001-menus"         "1001 Menus"                     None   
9           "/company/100du-tv"           "100du.tv"                     None   

  acquired_date  
0          None  
1          None  
2          None  
3          None  
4          None  
5          None  
6     "10/3/12"  
7          None  
8          None  
9          None

In [52]:
%%sql

SELECT companies.permalink AS companies_permalink,
    companies.name AS companies_name,
    acquisitions.company_permalink AS acquisitions_permalink,
    acquisitions.acquired_at AS acquired_date
FROM crunchbase_companies companies
    LEFT JOIN crunchbase_acquisitions acquisitions ON companies.permalink = acquisitions.company_permalink
    AND acquisitions.company_permalink != '"/company/1000memories"'
ORDER BY 1
LIMIT 10

companies_permalink       companies_name acquisitions_permalink  \
0            "/company/0-6-com"            "0-6.com"                   None   
1             "/company/0xdata"             "0xdata"                   None   
2      "/company/1-800-doctors"      "1-800-DOCTORS"                   None   
3                  "/company/1"           "One Inc."                   None   
4        "/company/10-20-media"        "10-20 Media"                   None   
5  "/company/1000jobboersen-de"  "1000jobboersen.de"                   None   
6       "/company/1000memories"       "1000memories"                   None   
7    "/company/1000museums-com"    "1000museums.com"                   None   
8         "/company/1001-menus"         "1001 Menus"                   None   
9           "/company/100du-tv"           "100du.tv"                   None   

  acquired_date  
0          None  
1          None  
2          None  
3          None  
4          None  
5          None  
6          None  
7          None  
8          None  
9          None

What's happening above is that the conditional statement AND... is evaluated before the join occurs. You can think of it as a WHERE clause that only applies to one of the tables. You can tell that this is only happening in one of the tables because the 1000memories permalink is still displayed in the column that pulls from the other table

If you move the same filter to the WHERE clause, you will notice that the filter happens after the tables are joined. The result is that the 1000memories row is joined onto the original table, but then it is filtered out entirely (in both tables) in the WHERE clause before displaying results.

In [51]:
%%sql

SELECT companies.permalink AS companies_permalink,
    companies.name AS companies_name,
    acquisitions.company_permalink AS acquisitions_permalink,
    acquisitions.acquired_at AS acquired_date
FROM crunchbase_companies companies
    LEFT JOIN crunchbase_acquisitions acquisitions ON companies.permalink = acquisitions.company_permalink
WHERE acquisitions.company_permalink != '"/company/1000memories"'
    OR acquisitions.company_permalink IS NULL
ORDER BY 1
LIMIT 10

companies_permalink       companies_name acquisitions_permalink  \
0            "/company/0-6-com"            "0-6.com"                   None   
1             "/company/0xdata"             "0xdata"                   None   
2      "/company/1-800-doctors"      "1-800-DOCTORS"                   None   
3                  "/company/1"           "One Inc."                   None   
4        "/company/10-20-media"        "10-20 Media"                   None   
5  "/company/1000jobboersen-de"  "1000jobboersen.de"                   None   
6    "/company/1000museums-com"    "1000museums.com"                   None   
7         "/company/1001-menus"         "1001 Menus"                   None   
8           "/company/100du-tv"           "100du.tv"                   None   
9           "/company/100e-com"           "100e.com"                   None   

  acquired_date  
0          None  
1          None  
2          None  
3          None  
4          None  
5          None  
6          None  
7          None  
8          None  
9          None

You can see that the 1000memories line is not returned (it would have been between the two highlighted lines below). Also note that filtering in the WHERE clause can also filter null values, so we added an extra line to make sure to include the nulls.

## SQL SELF JOIN

The SELF JOIN allows you to join a table to itself. This implies that each row of the table is combined with itself and with every other row of the table. The SELF JOIN can be viewed as a join of two copies of the same table. The table is not actually copied, but SQL performs the command as though it were. This is accomplished by using table name aliases to give each instance of the table a separate name. It is most useful for extracting hierarchical data or comparing rows within the same table.

## SQL CROSS JOIN

The CROSS JOIN command in SQL, also known as a cartesian join, returns all combinations of rows from each table. Envision that you need to find all combinations of size and color. In that case, a CROSS JOIN will be an asset. Note, that this join does not need any condition to join two tables. In fact, CROSS JOIN joins every row from the first table with every row from the second table and its result comprises all combinations of records in two tables.

In [ ]:
Select * from table_1 cross join table_2;  

In [ ]:
Select column1, column2, column3 FROM table_1, table_2;  

## SQL UNION

In [55]:
%%sql

SELECT *
FROM crunchbase_acquisitions
LIMIT 10

company_permalink  \
0                                 "/company/waywire"   
1                     "/company/1-nation-technology"   
2  "/company/1-stop-financial-service-centers-of-...   
3                        "/company/1-800-contacts-2"   
4                            "/company/1000memories"   
5                                  "/company/10best"   
6                                  "/company/10east"   
7                              "/company/10k-wizard"   
8                           "/company/123linkit-com"   
9                               "/company/123people"   

                                    company_name company_category_code  \
0                                     "#waywire"                "news"   
1                          "1 Nation Technology"                         
2  "1 Stop Financial Service Centers of America"                         
3                               "1-800 Contacts"                         
4                                 "1000memories"                 "web"   
5                                       "10best"                 "web"   
6                                       "10East"     "network_hosting"   
7                                   "10K Wizard"                         
8                                "123LinkIt.com"                         
9                                    "123people"              "search"   

  company_country_code company_state_code  company_region     company_city  \
0                "USA"               "NY"      "New York"       "New York"   
1                                               "unknown"                    
2                "USA"               "TX"        "Austin"     "Round Rock"   
3                                               "unknown"                    
4                "USA"               "CA"        "SF Bay"  "San Francisco"   
5                "USA"               "SC"    "Greenville"     "Greenville"   
6                "USA"               "FL"  "Jacksonville"   "Jacksonville"   
7                                               "unknown"                    
8                                               "unknown"                    
9                "AUT"                           "Vienna"         "Vienna"   

                 acquirer_permalink             acquirer_name  \
0                "/company/magnify"                 "Magnify"   
1                 "/company/vology"                  "Vology"   
2         "/company/confie-seguros"          "Confie Seguros"   
3  "/company/thomas-h-lee-partners"  "Thomas H. Lee Partners"   
4           "/company/ancestry-com"                "Ancestry"   
5             "/company/nile-guide"               "NileGuide"   
6     "/company/railcar-management"      "Railcar Management"   
7            "/company/morningstar"             "Morningstar"   
8                "/company/netline"                 "NetLine"   
9            "/company/pagesjaunes"             "PagesJaunes"   

  acquirer_category_code  ... acquirer_state_code   acquirer_region  \
0          "games_video"  ...                "NY"        "New York"   
1                "other"  ...                             "unknown"   
2           "enterprise"  ...                "CA"     "Los Angeles"   
3                         ...                "MA"          "Boston"   
4            "ecommerce"  ...                "UT"  "Salt Lake City"   
5               "social"  ...                "CA"          "SF Bay"   
6           "enterprise"  ...                             "unknown"   
7             "software"  ...                "IL"         "Chicago"   
8          "advertising"  ...                "CA"          "SF Bay"   
9          "advertising"  ...                               "Paris"   

     acquirer_city acquired_at acquired_month acquired_quarter acquired_year  \
0       "New York"  "10/17/13"      "2013-10"        "2013-Q4"        2013.0   
1                     "1/1/06"      "2006-01"        "2006-Q1"        2006.0   
2     

In [58]:
%%sql

SELECT *
FROM crunchbase_acquisitions
WHERE company_permalink = '"/company/waywire"'
UNION
SELECT *
FROM crunchbase_acquisitions
WHERE company_permalink = '"/company/80-20"'
LIMIT 10

company_permalink company_name company_category_code company_country_code  \
0  "/company/waywire"   "#waywire"                "news"                "USA"   
1    "/company/80-20"      "80/20"            "hardware"                "USA"   

  company_state_code company_region company_city  acquirer_permalink  \
0               "NY"     "New York"   "New York"  "/company/magnify"   
1               "NY"     "New York"   "New York"   "/company/square"   

  acquirer_name acquirer_category_code  ... acquirer_state_code  \
0     "Magnify"          "games_video"  ...                "NY"   
1      "Square"               "mobile"  ...                "CA"   

  acquirer_region    acquirer_city acquired_at acquired_month  \
0      "New York"       "New York"  "10/17/13"      "2013-10"   
1        "SF Bay"  "San Francisco"   "10/1/12"      "2012-10"   

  acquired_quarter acquired_year  price_amount  price_currency_code  id  
0        "2013-Q4"        2013.0           0.0                "USD"   1  
1        "2012-Q4"        2012.0           0.0                "USD"  70  

[2 rows x 21 columns]

SQL Joins with Comparison Operators

In [60]:
%%sql

SELECT companies.permalink,
    companies.name,
    companies.status,
    COUNT(investments.investor_permalink) AS investors
FROM crunchbase_companies companies
    LEFT JOIN crunchbase_investments_part1 investments 
        ON companies.permalink = investments.company_permalink
    AND investments.funded_year > companies.founded_year + 5
GROUP BY 1,
    2,
    3
LIMIT 10

permalink                        name  \
0              "/company/a-list-games"             "[a]list games"   
1                  "/company/1o1media"                  "1o1Media"   
2                   "/company/1stdibs"                   "1stdibs"   
3                     "/company/4info"                     "4INFO"   
4                     "/company/500px"                     "500px"   
5                   "/company/7-bites"                   "7-bites"   
6                    "/company/8hands"                    "8hands"   
7  "/company/91-financial-supermarket"  "91 Financial Supermarket"   
8           "/company/a-curated-world"           "A Curated World"   
9               "/company/a123systems"              "A123 Systems"   

        status  investors  
0  "operating"          1  
1  "operating"          1  
2  "operating"          4  
3  "operating"          5  
4  "operating"          7  
5  "operating"          2  
6     "closed"          1  
7  "operating"          2  
8  "operating"          1  
9        "ipo"          8

In [61]:
%%sql

SELECT companies.permalink,
    companies.name,
    companies.status,
    COUNT(investments.investor_permalink) AS investors
FROM crunchbase_companies companies
    LEFT JOIN crunchbase_investments_part1 investments 
        ON companies.permalink = investments.company_permalink
WHERE investments.funded_year > companies.founded_year + 5
GROUP BY 1,
    2,
    3
LIMIT 10

permalink  \
0                         "/company/advercar"   
1                  "/company/azure-solutions"   
2           "/company/alvine-pharmaceuticals"   
3                 "/company/avaxia-biologics"   
4  "/company/airband-communications-holdings"   
5                           "/company/apigee"   
6                          "/company/1stdibs"   
7               "/company/alti-semiconductor"   
8                     "/company/ancestry-com"   
9                         "/company/bambuser"   

                                name       status  investors  
0                         "AdverCar"  "operating"          4  
1                  "Azure Solutions"   "acquired"          4  
2           "Alvine Pharmaceuticals"  "operating"          1  
3                 "Avaxia Biologics"  "operating"         11  
4  "Airband Communications Holdings"  "operating"          5  
5                           "Apigee"  "operating"          8  
6                          "1stdibs"  "operating"          4  
7               "Alti Semiconductor"  "operating"          1  
8                         "Ancestry"        "ipo"          3  
9                         "Bambuser"  "operating"          1

SQL Joins on Multiple Keys

In [62]:
%%sql

SELECT companies.permalink,
    companies.name,
    investments.company_name,
    investments.company_permalink
FROM crunchbase_companies companies
    LEFT JOIN crunchbase_investments_part1 investments 
        ON companies.permalink = investments.company_permalink
    AND companies.name = investments.company_name
LIMIT 10

permalink                 name         company_name  \
0               "/company/8868"               "8868"               "8868"   
1       "/company/a-list-games"      "[a]list games"      "[a]list games"   
2  "/company/1366-technologies"  "1366 Technologies"  "1366 Technologies"   
3           "/company/1o1media"           "1o1Media"           "1o1Media"   
4              "/company/1ring"              "1Ring"              "1Ring"   
5            "/company/1stdibs"            "1stdibs"            "1stdibs"   
6         "/company/1stgig-com"         "1stGig.com"         "1stGig.com"   
7         "/company/2080-media"         "2080 Media"         "2080 Media"   
8            "/company/23andme"            "23andMe"            "23andMe"   
9            "/company/23andme"            "23andMe"            "23andMe"   

              company_permalink  
0               "/company/8868"  
1       "/company/a-list-games"  
2  "/company/1366-technologies"  
3           "/company/1o1media"  
4              "/company/1ring"  
5            "/company/1stdibs"  
6         "/company/1stgig-com"  
7         "/company/2080-media"  
8            "/company/23andme"  
9            "/company/23andme"

SQL Self Joins

Sometimes it can be useful to join a table to itself. Let’s say you wanted to identify companies that received an investment from Great Britain following an investment from Japan.

In [64]:
%%sql

SELECT DISTINCT japan_investments.company_name,
    japan_investments.company_permalink
FROM crunchbase_investments_part1 japan_investments
    JOIN crunchbase_investments_part1 gb_investments 
        ON japan_investments.company_name = gb_investments.company_name
    AND gb_investments.investor_country_code = '"GBR"'
    AND gb_investments.funded_at > japan_investments.funded_at
WHERE japan_investments.investor_country_code = '"JPN"'
ORDER BY 1
LIMIT 10

company_name                   company_permalink
0                   "Avnera"                   "/company/avnera"
1               "Crystal IS"               "/company/crystal-is"
2  "Dicerna Pharmaceuticals"  "/company/dicerna-pharmaceuticals"
3                  "Enforta"                  "/company/enforta"
4                   "Eye-Fi"                   "/company/eye-fi"
5                  "Fab.com"                  "/company/fab-com"
6   "ForeScout Technologies"                "/company/forescout"
7                    "Hailo"                    "/company/hailo"
8            "IGA Worldwide"             "/company/igaworldwide"
9                 "Memolane"                 "/company/memolane"

Note how the same table can easily be referenced multiple times using different aliases—in this case, japan_investments and gb_investments.